### PR TITLE
Scene Graph Integration

### PR DESCRIPTION
**Pull request by benptc**
Very large PR fully swapping out the way positions of everything in AR are calculated with more optimized and standardized scene graph. App runs much more smoothly now, and this data model will make spatial features easier to implement.

**Implementation:**
- **src/sceneGraph/sceneNode.js**: a class for managing the position of one entity in space. It has a *localMatrix* (where it is relative to parent) and a *worldMatrix* (resulting position in world coordinates). Results are automatically cached for better performance and invalidated when needed.
- **src/sceneGraph/index.js**: maintains a data structure of SceneNodes to represent Objects, Frames, Nodes, the Camera, the Groundplane, and optional additional elements. Provides a number of APIs for updating the positions of things and querying the scene graph for information.

**Integrations:**
- the main render loop is updated to use the scene graph instead of ad-hoc computations
- all repositioning methods (dragging, scaling, etc) are updated
- world objects and groundplane are updated
- (also includes unrelated optimizations for groundplane and a button to recalculate it)
- anchor objects are updated
- memories are updated (also improved pocket UX while saving memory)
- spatial questions are updated
- frame APIs (e.g. for three.js) are updated
- all sorts of distance calculations are updated (e.g. fading out, viewport culling)

I still have a long list of little things related to this (some are noted with todo comments), but it's reached a point where it works again and can be merged to master.

### CODE REVIEW

**Review by afustePTC — APPROVED** [review]
Great

File: src/gui/ar/positioning.js [hunk 1]
```diff
@@ -137,28 +199,42 @@ realityEditor.gui.ar.positioning.scaleVehicle = function(activeVehicle, centerTo
...
  */
-realityEditor.gui.ar.positioning.moveVehicleToScreenCoordinate = function(activeVehicle, screenX, screenY, useTouchOffset) {
+realityEditor.gui.ar.positioning.moveVehicleToScreenCoordinate = function(activeVehicle, screenX, screenY, useTouchOffset, isContinuous) {
     
-    var results;
+    var newPosition;
     try {
-        results = realityEditor.gui.ar.utilities.screenCoordinatesToMatrixXY(activeVehicle, screenX, screenY, true);
+        // set dummy div transform to iframe without x,y,scale
+        if (this.useWebkitForProjectedCoordinaates) {
```
Comment (hobinjk-ptc) [inline src/gui/ar/positioning.js:210]: typo

File: src/sceneGraph/index.js [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+            // let _object = realityEditor.getObject(objectId);
+            sceneGraph[thisNodeId] = sceneNodeRotateX;
+        }
+
+        sceneNodeRotateX.setParent(sceneNodeObject);
+        console.log('SceneGraph: added rotateX to object ' + objectId);
+
+        // image target objects require one coordinate system rotation. ground plane requires another.
+        if (groundPlaneVariation) {
+            sceneNodeRotateX.setLocalMatrix( makeGroundPlaneRotationX(-(Math.PI/2)) );
```
Comment (hobinjk-ptc) [inline src/sceneGraph/index.js:507]: nit: Parentheses are a little weird here

**Review by hobinjk-ptc — APPROVED** [review]
Excellent work! 1 typo inline

File: src/device/index.js [hunk 3]
```diff
@@ -1324,58 +1320,42 @@ realityEditor.device.checkIfFramePulledIntoUnconstrained = function(activeVehicl
...
+        let dx = camPos.x - this.editingState.initialCameraPosition.x;
+        let dy = camPos.y - this.editingState.initialCameraPosition.y;
+        let dz = camPos.z - this.editingState.initialCameraPosition.z;
 
-        // the first time, detect how far you are from the element and use that as a baseline
-        if (!this.editingState.unconstrainedOffset) {
-            this.editingState.unconstrainedOffset = distanceToObject;
+        let cameraMoveDistance = Math.sqrt(dx * dx + dy * dy + dz * dz);
+        
+        // TODO ben: for frames on screen object, if direction is towards screen then push into screen instead
```
[the remark this answers is not in the document]
Comment (benptc) [reply src/device/index.js:1336]: One of my next projects is getting the screen<>AR integration working again. Pushing doesn't work (as seen here). Pulling also doesn't work (gui/screenExtension.js)

File: src/device/onLoad.js [hunk 4]
```diff
@@ -165,6 +165,19 @@ realityEditor.device.onload = function () {
         }
     }).moveToDevelopMenu();
 
+    // set up the global canvas for drawing the links
+    globalCanvas.canvas = document.getElementById('canvas');
+    globalCanvas.canvas.width = globalStates.height; // TODO: fix width vs height mismatch once and for all
```
[the remark this answers is not in the document]
Comment (benptc) [reply src/device/onLoad.js:170]: Indeed

File: src/gui/ar/draw.js [hunk 5]
```diff
@@ -691,21 +608,23 @@ realityEditor.gui.ar.draw.update = function (visibleObjects) {
...
     
-    // todo: only do this if there are any closestObjectListeners
-    var newClosestObject = realityEditor.gui.ar.getClosestObject()[0];
-    if (newClosestObject !== this.currentClosestObject) {
-        this.closestObjectListeners.forEach(function(callback) {
-            callback(this.currentClosestObject, newClosestObject); 
-        }.bind(this));
-        this.currentClosestObject = newClosestObject;
+    // todo ben: only do this if there are any closestObjectListeners
+    if (this.closestObjectListeners.length > 0 && this.isLowFrequencyUpdateFrame) {
```
[the remark this answers is not in the document]
Comment (benptc) [reply src/gui/ar/draw.js:612]: ah, forgot to remove the comment 👍

File: src/gui/ar/positioning.js (hunk 1 above)
Comment (benptc) [reply src/gui/ar/positioning.js:210]: good catch. my MacBook keyboard is a bit messed up and double-types letters with alarming frequency.

File: src/sceneGraph/index.js (hunk 2 above)
Comment (benptc) [reply src/sceneGraph/index.js]: Thank you for such a thorough review!

**Pull request merged**